crypto/evp/ctrl_params_translate.c

```diff
@@ -693,6 +693,9 @@ static int default_fixup_args(enum state state,
                                    translation->param_data_type);
                     return 0;
                 }
+            } else if (state == PRE_PARAMS_TO_CTRL && ctx->action_type == GET) {
+                /* Assign output */
+                ctx->p2 = ctx->params->data;
```

Comment:
This adds a few necessary checks, please try it:

```suggestion
            } else if (state == PRE_PARAMS_TO_CTRL && ctx->action_type == GET
                       && translation->param_data_type == OSSL_PARAM_OCTET_PTR) {
                /* Assign output */
                if (!OSSL_PARAM_get_octet_ptr(ctx->params, &ctx->p2, &ctx->sz))
                    return 0;              
```

Reply:
After calling `OSSL_PARAM_get_octet_ptr`, `ctx->p2` remains NULL.
To mitigate the potential side effects in other functions, I have refactored the code.

Reply:
> After calling `OSSL_PARAM_get_octet_ptr`, `ctx->p2` remains NULL. To mitigate the potential side effects in other functions, I have refactored the code.

This is weird. I thought it should work. Using special fixup function is not right as there are other parameters that should be handled in the exactly same way and currently aren't working.

Reply:
Please try this instead:

```diff
diff --git a/crypto/evp/ctrl_params_translate.c b/crypto/evp/ctrl_params_translate.c
index 480d48429b..313f7d2b72 100644
--- a/crypto/evp/ctrl_params_translate.c
+++ b/crypto/evp/ctrl_params_translate.c
@@ -647,6 +647,9 @@ static int default_fixup_args(enum state state,
                                    translation->param_data_type);
                     return 0;
                 }
+            } else if (state == PRE_PARAMS_TO_CTRL && ctx->action_type == GET) {
+                if (translation->param_data_type == OSSL_PARAM_OCTET_PTR)
+                    ctx->p2 = &ctx->bufp;
             } else if ((state == POST_PARAMS_TO_CTRL || state == PKEY)
                        && ctx->action_type == GET) {
                 /* For the POST state, only getting needs some work to be done */
```

Reply:
That works. Now, I need to find a suitable place to add an extra dereference to obtain the correct value of the label. Specifically, I have to place `ctx->p2 = *(unsigned char**)ctx->p2` within the `POST_PARAMS_TO_CTRL` step. However, it does not seem to fit well.

Reply:
> That works. Now, I need to find a suitable place to add an extra dereference to obtain the correct value of the label. Specifically, I have to place `ctx->p2 = *(unsigned char**)ctx->p2` within the `POST_PARAMS_TO_CTRL` step. However, it does not seem to fit well.

You don't. This line handles it https://github.com/openssl/openssl/pull/21401/files#diff-910515c87ac6c81bba3f87a197f149cdcc050330964f3ea26877217df4315e0eL683

Reply:
> 

We need one more dereference, as without it, I am getting random data in the current state.

Reply:
Actually sorry, the line 686 is the right one and the size handling is not correct there either. What about: `return OSSL_PARAM_set_octet_ptr(ctx->params, *((void **)ctx->p2), ret);`